cwl/src/main/scala/wdl4s/cwl/Workflow.scala

```diff
@@ -0,0 +1,134 @@
+      cwlTypeForInputParameter(input) map cwlTypeToWdlType
+    }
+
+    val typeMap: WdlTypeMap =
+      outputsTypeMap ++
+        // Note this is only looking at the workflow inputs and not recursing into steps, because our current thinking
+        // is that in CWL graph inputs can only be defined at the workflow level.  It's possible that's not actually
+        // correct, but that's the assumption being made here.
+        inputs.toList.flatMap { i =>
+          wdlTypeForInputParameter(i).map(i.id -> _).toList
+        }.toMap
+
+    val graphFromInputs: Set[RequiredGraphInputNode] = inputs.map {
+      input => RequiredGraphInputNode(input.id, wdlTypeForInputParameter(input).get)
+    }.toSet
+
+    val workflowInputs: Map[String, GraphNodeOutputPort] =
+      graphFromInputs.map {
+        workflowInput =>
+          workflowInput.name -> GraphNodeOutputPort(workflowInput.name, workflowInput.womType, workflowInput)
+      }.toMap
+
+    val graphFromSteps =
+      steps.
+        toList.
+        foldLeft(Set.empty[GraphNode])(
+          (nodes, step) => step.callWithInputs(typeMap,  this, nodes, workflowInputs))
+
+    val graphFromOutputs: ErrorOr[Set[GraphNode]] =
+      outputs.toList.traverse[ErrorOr, GraphNode] {
+        output =>
+
+          val wdlType = cwlTypeToWdlType(output.`type`.flatMap(_.select[CwlType]).get)
+
+          def lookupOutputSource(source: String): ErrorOr[OutputPort] =
+            graphFromSteps.
+              flatMap(_.outputPorts).
+              find(_.name == source).
+              toValidNel(s"unable to find upstream port corresponding to $source")
+
+          lookupOutputSource(output.outputSource.flatMap(_.select[String]).get).
+            map(PortBasedGraphOutputNode(output.id, wdlType, _))
+      }.map(_.toSet)
+
+    graphFromOutputs.flatMap(outputs => Graph.validateAndConstruct(graphFromSteps ++ graphFromInputs ++ outputs))
+  }
+
+  def womDefinition: ErrorOr[WorkflowDefinition] = {
+    val name: String = "workflow Id"
+    val meta: Map[String, String] = Map.empty
+    val paramMeta: Map[String, String] = Map.empty
+    val declarations: List[(String, WomExpression)] = List.empty
+
+    womGraph.map(graph =>
+      WorkflowDefinition(
+        name,
+        graph,
+        meta,
+        paramMeta,
+        declarations
+      )
+    )
+  }
+
+  def asCwl = Coproduct[Cwl](this)
+}
+object Workflow {
+
+  //run this thing and give me an output
+  def minimal(wfs: WorkflowStep) = {
+  }
+  def validate(
+    name: String,
+    steps: Set[WorkflowStep],
+    version: Option[CwlVersion] = None,
+    inputs: Array[InputParameter] = Array.empty,
+    outputs: Array[WorkflowOutputParameter] = Array.empty): Either[NonEmptyList[String], Cwl] = {
+
+
+    //does no checking at all
+    val wf = Workflow(steps = steps.toArray)
+    Right(Coproduct[Cwl](wf))
+  }
+
+  val `class`: Witness.`"Workflow"`.T = "Workflow".narrow
```

Comment:
* move this to Constructor type and get rid of this as it's not being properly enforced.
 * make Workflow constructor private and direct people to use constructor above, so that awkwardness around providing this value is avoided.